caproto/_data.py

```diff
@@ -751,13 +758,69 @@ def __init__(self, *, value, units='',
             if not self.lower_ctrl_limit <= val <= self.upper_ctrl_limit:
                 raise CannotExceedLimits(
                     f"Cannot write data {val}. Limits are set to "
-                    f"{self.lower_ctrl_limit} and {self.upper_ctrl_limit}.")
-        if self.lower_warning_limit != self.upper_warning_limit:
-            if not self.lower_ctrl_limit <= val <= self.upper_warning_limit:
-                warnings.warn(
-                    f"Writing {val} outside warning limits which are are "
-                    f"set to {self.lower_ctrl_limit} and "
-                    f"{self.upper_warning_limit}.")
+                    f"{self.lower_ctrl_limit} and {self.upper_ctrl_limit}."
+                )
+
+        async def limit_checker(
+                value,
+                lo_attr, hi_attr,
+                lo_status, hi_status,
+                lo_severity_attr,
+                hi_severity_attr,
+                dflt_lo_severity,
+                dflt_hi_severity):
+
+            def limit_getter(limit_attr, severity_attr, dflt_severity):
+                sev = dflt_severity
+                limit = getattr(self, limit_attr)
+
+                sev_prop = getattr(
+                    getattr(self, 'field_inst', None),
+                    severity_attr, None)
+                if sev_prop is not None:
+                    sev = sev_prop.enum_strings.index(sev_prop.value)
+                return limit, AlarmSeverity(sev)
+
+            lo_limit, lo_severity = limit_getter(
+                lo_attr, lo_severity_attr, dflt_lo_severity)
+            hi_limit, hi_severity = limit_getter(
+                hi_attr, hi_severity_attr, dflt_hi_severity)
+            if lo_limit != hi_limit:
+                if value <= lo_limit:
+                    await self.alarm.write(status=lo_status,
+                                           severity=lo_severity)
+                    return False
+                elif hi_limit <= value:
+                    await self.alarm.write(status=hi_status,
+                                           severity=hi_severity)
+                    return False
+
+            await self.alarm.write(status=AlarmStatus.NO_ALARM,
+                                   severity=AlarmSeverity.NO_ALARM)
+            return True
+
+        # this is HIHI and LOLO limits
+        ok = await limit_checker(val,
```

Comment:
Can you pull these out into not-async functions and `await` on `alarm.write` only once?

At some point, we should start worrying more about where we `await` unnecessarily, giving the event loop a chance to switch to some other task.

Reply:
Isn't more await better to give it more chances to multi-task?

Reply:
I'd argue that we want more atomicity in simple operations such as this - there is no good reason to task switch out of this.

Thoughts @danielballan ?

Reply:
> I'd argue that we want more atomicity in simple operations such as this - there is no good reason to task switch out of this.

Agreed. If there is a series of operations that we know to be fast (i.e. no I/O) then there is no reason to switch tasks in the middle. That just adds overhead.

Seems like we are in agreement now, in any case.